src/taco-remote-lib/common/builder.ts

```diff
@@ -123,13 +124,98 @@ class Builder {
+     * Returns true if the platform was added, false if not.
+     */
+    private ensurePlatformAdded(): Q.Promise<boolean> {
         if (!fs.existsSync(path.join("platforms", this.currentBuild.buildPlatform))) {
             Logger.log("cordova platform add " + this.currentBuild.buildPlatform);
+
             // Note that "cordova platform add" eventually calls "cordova prepare" internally, which is why we don't invoke prepare ourselves when we add the platform.
-            return this.cordova.raw.platform("add", this.currentBuild.buildPlatform);
-        } else {
-            return this.update_platform();
+            return this.cordova.raw.platform("add", this.currentBuild.buildPlatform).then(() => {
+                return Q(true);
+            });
+        }
+
+        var xmlVersion: string = this.getConfigXmlPlatformVersion();
+
+        return this.getCurrentPlatformVersion().then((installedVersion: string) => {
+            if (xmlVersion) {
+                if (semver.neq(xmlVersion, installedVersion)) {
+                    // There is a version of the platform specified in xonfig.xml, and it differs from the installed version
+                    return true;
+                } else {
+                    // The version specified in config.xml is already installed, we don't need to re-add the platform
+                    return false;
+                }
+            } else {
+                // Check if the Cordova version has changed
+                if (this.currentBuild["previousvcordova"] && semver.neq(this.currentBuild["vcordova"], this.currentBuild["previousvcordova"])) {
+                    // Cordova version has changed
+                    return true;
+                }
+
+                // Cordova version hasn't changed
+                return false;
+            }
+        }, (err) => {
+            // If we couldn't extract the platform's version, then remove and re-add it anyway, to be safe
+            return true;
+        }).then((mustReAdd: boolean) => {
+            if (!mustReAdd) {
+                return Q(false);
+            } else {
+                Logger.log("cordova platform remove " + this.currentBuild.buildPlatform);
+
+                return this.cordova.raw.platform("remove", this.currentBuild.buildPlatform).then(() => {
+                    Logger.log("cordova platform add " + this.currentBuild.buildPlatform);
+
+                    return this.cordova.raw.platform("add", this.currentBuild.buildPlatform);
+                }).then(() => {
+                    return Q(true);
+                });
+            }
+        });
+    }
+
+    private getConfigXmlPlatformVersion(): string {
+        var engines: {[key: string]: string} = utils.CordovaConfig.getCordovaConfig(this.currentBuild.appDir).engines();
```

Comment:
I just went with our convention of always specifying the types, but I can remove it